### PR TITLE
change-api-products

### PR DESCRIPTION
**Pull request by ghost**
## Description
Change api products
## Reference Requirements
- Task ID: TI-1

## Reference Isuees
#{issue_id}

## Migrations
NO

## Impacted Areas in Application
List general components of the application that this PR will affect:

## SQL query and `explain`
List of all queries and attach the result of each explaining query (in text, screenshot...)

- explain select * from `categories` where `categories`.`id` in (1, 4, 5, 6) and `categories`.`deleted_at` is null \G;
*************************** 1. row ***************************
           id: 1
  select_type: SIMPLE
        table: categories
   partitions: NULL
         type: range
possible_keys: PRIMARY
          key: PRIMARY
      key_len: 4
          ref: NULL
         rows: 4
     filtered: 10.00
        Extra: Using where
1 row in set, 1 warning (0.00 sec)

- explain select * from `products` where exists (select * from `categories` where `products`.`category_id` = `categories`.`id` and `id` in (1, 3, 4, 5, 6, 11, 12) and `categories`.`deleted_at` is null) and `products`.`deleted_at` is null limit 15 offset 0 \G
*************************** 1. row ***************************
           id: 1
  select_type: PRIMARY
        table: products
   partitions: NULL
         type: ALL
possible_keys: NULL
          key: NULL
      key_len: NULL
          ref: NULL
         rows: 20
     filtered: 10.00
        Extra: Using where
*************************** 2. row ***************************
           id: 2
  select_type: DEPENDENT SUBQUERY
        table: categories
   partitions: NULL
         type: eq_ref
possible_keys: PRIMARY
          key: PRIMARY
      key_len: 4
          ref: food_drink.products.category_id
         rows: 1
     filtered: 10.00
        Extra: Using where
2 rows in set, 2 warnings (0.00 sec)

- explain select count(*) as aggregate from `products` where exists (select * from `categories` where `products`.`category_id` = `categories`.`id` and `id` in (1, 3, 4, 5, 6, 11, 12) and `categories`.`deleted_at` is null) and `products`.`deleted_at` is null \G
*************************** 1. row ***************************
           id: 1
  select_type: PRIMARY
        table: products
   partitions: NULL
         type: ALL
possible_keys: NULL
          key: NULL
      key_len: NULL
          ref: NULL
         rows: 20
     filtered: 10.00
        Extra: Using where
*************************** 2. row ***************************
           id: 2
  select_type: DEPENDENT SUBQUERY
        table: categories
   partitions: NULL
         type: eq_ref
possible_keys: PRIMARY
          key: PRIMARY
      key_len: 4
          ref: food_drink.products.category_id
         rows: 1
     filtered: 10.00
        Extra: Using where
2 rows in set, 2 warnings (0.00 sec)

- explain select * from `categories` where `categories`.`parent_id` in (3, 11, 12) and `categories`.`deleted_at` is null \G
*************************** 1. row ***************************
           id: 1
  select_type: SIMPLE
        table: categories
   partitions: NULL
         type: ALL
possible_keys: NULL
          key: NULL
      key_len: NULL
          ref: NULL
         rows: 12
     filtered: 8.33
        Extra: Using where
1 row in set, 1 warning (0.00 sec)

- explain select * from `categories` where `categories`.`parent_id` = 1 and `categories`.`parent_id` is not null and `categories`.`deleted_at` is null \G
*************************** 1. row ***************************
           id: 1
  select_type: SIMPLE
        table: categories
   partitions: NULL
         type: ALL
possible_keys: NULL
          key: NULL
      key_len: NULL
          ref: NULL
         rows: 12
     filtered: 8.33
        Extra: Using where
1 row in set, 1 warning (0.00 sec)

- explain select * from `categories` where `id` = 1 and `categories`.`deleted_at` is null limit 1 \G
*************************** 1. row ***************************
           id: 1
  select_type: SIMPLE
        table: categories
   partitions: NULL
         type: const
possible_keys: PRIMARY
          key: PRIMARY
      key_len: 4
          ref: const
         rows: 1
     filtered: 100.00
        Extra: NULL
1 row in set, 1 warning (0.00 sec)

### CODE REVIEW

File: app/Http/Controllers/Api/ProductController.php [hunk 1]
```diff
@@ -22,6 +25,10 @@ class ProductController extends ApiController
      */
     public function index(SortApiProductRequest $request)
     {
+        $categoriesId = 0;
+        if ($request->category) {
+            $categoriesId = $this->arrayCategoryId($request->category);
```
Comment (at-thaitran) [inline app/Http/Controllers/Api/ProductController.php]: không cần ì chỗ ni, bỏ vô đoạn `when(isset($request->category)..` luôn

File: app/Traits/FillterTrait.php [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+    public function arrayCategoryId($id)
+    {
+        $categoryAll = Category::where('id', $id)->first();
+        $categoryRelation = $categoryAll->children()->with('children')->get();
+        $categoriesId = array();
+        array_push($categoriesId, $id);
+        foreach ($categoryRelation as $key => $value) {
+            array_push($categoriesId, $value->id);
+            foreach ($value->children as $key => $value) {
+                    array_push($categoriesId, $value->id);
```
Comment (at-thaitran) [inline app/Traits/FillterTrait.php]: dư tab

**Comment by at-thaitran** [comment]
LGTM

File: app/Traits/FillterTrait.php [hunk 3]
```diff
@@ -0,0 +1,30 @@
+<?php
+
+namespace App\Traits;
+
+use App\Category;
+
+trait FillterTrait
```
Comment (at-xinhnguyen) [inline app/Traits/FillterTrait.php]: thừa chữ `l` em `filter`

File: app/Traits/FillterTrait.php [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+    /**
+     * Make to array category id
+     *
+     * @param id $id id of category
+     *
+     * @return array category id
+     */
+    public function arrayCategoryId($id)
+    {
+        $categoryAll = Category::where('id', $id)->first();
```
Comment (at-xinhnguyen) [inline app/Traits/FillterTrait.php]: em đặt tên sao k đúng ý nghĩa vậy=> đã where theo `id`, `first` lại đặt tên là `all`

File: app/Http/Controllers/Api/ProductController.php [hunk 5]
```diff
@@ -35,7 +38,7 @@ public function index(SortApiProductRequest $request)
                     })
                     ->when(isset($request->category), function ($query) use ($request) {
                         return $query->whereHas('category', function ($query) use ($request) {
```
Comment (at-xinhnguyen) [inline app/Http/Controllers/Api/ProductController.php]: chỗ này sao lại cần `whereHas` làm gì em

**Comment by at-xinhnguyen** [comment]
Log hết câu query lên đi em

File: app/Http/Controllers/Api/ProductController.php (hunk 5 above)
Comment (at-sonle) [reply app/Http/Controllers/Api/ProductController.php]: tại query tới bảng category

**Pull request merged**